src/ledger/LedgerManagerImpl.cpp

```diff
@@ -579,12 +589,17 @@ LedgerManagerImpl::closeLedger(LedgerCloseData const& ledgerData)
     // LedgerHeader, we optionally collect an even-more-fine-grained record of
     // the ledger entries modified by each tx during tx processing in a
     // LedgerCloseMeta, for streaming to attached clients (typically: horizon).
-    std::unique_ptr<LedgerCloseMeta> ledgerCloseMeta;
     if (mMetaStream)
     {
-        ledgerCloseMeta = std::make_unique<LedgerCloseMeta>();
-        ledgerCloseMeta->v0().txProcessing.reserve(txSet->sizeTx());
-        txSet->toXDR(ledgerCloseMeta->v0().txSet);
+        if (mNextMetaToEmit)
+        {
+            releaseAssert(mNextMetaToEmit->v0().ledgerHeader.hash ==
+                          getLastClosedLedgerHeader().hash);
+            emitNextMeta();
+        }
+        mNextMetaToEmit = std::make_unique<LedgerCloseMeta>();
```

Comment:
This is not exception safe:
it's entirely possible that `LedgerManagerImpl::closeLedger` gets called and throws after this point - I think if this happens the caller will just retry (if this is not possible right now, writing the code this way is a clear footgun if we ever want to change that).

So what we want to do here is consume `mNextMetaToEmit` like it is now (consensus was reached so we can emit the meta from the previous ledger), and continue to use a local variable `ledgerCloseMeta` like before, and only at the end move it into `mNextMetaToEmit` when it is safe to do so.

Reply:
Fixed.